Configurable version increments per Type

ccopsey opened the pull request:
I expect that this PR will be declined as it takes a slightly different view of the ConventionalCommit spec, and actively deviates from the ConventionalChangelog spec, however I would appreciate feedback if possible.

I like convco very much.  However, to be of real value to me I need to be able increment either the major, minor or patch via custom types, not constrained to `feat` and `fix`.  This is allowed by the spec, but not mandated.  This PR removes `conventional::Type` preferring instead for every type to be a `String`.  It then adds a new `Increment` config Type field which allows the specification of which value to increment when that type is encountered in the git history.

hdevalke commented:
Thank you for the PR. Convco already supports custom types https://github.com/convco/convco/blob/master/src/conventional/commits.rs#L20,  but not an increment behavior. I will have a look to your PR later  in more details

hdevalke commented on src/conventional/config.rs:
```diff
@@ -8,9 +8,17 @@ use url::Url;
 use crate::{error::Error, git::GitHelper};
 
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub(crate) enum Increment {
```
This is missing a None variant

hdevalke commented on src/conventional/config.rs:
```diff
@@ -135,51 +143,61 @@ fn default_types() -> Vec<Type> {
         },
         Type {
             r#type: "build".into(),
+            increment: Increment::Patch,
```
These none fix/feature types should default to the `Increment::None`. Otherwise this is a breaking change.

ccopsey replied:
I see what you are saying and will update.  Theoretically though, should any new release, even if the interim commits are `chore` or `build` or something else minor, not expect at least a patch increment?

hdevalke replied:
> Additional types are not mandated by the Conventional Commits specification, and have no implicit effect in Semantic Versioning (unless they include a BREAKING CHANGE). 

Unless they contain a breaking change there should not be an implicit effect. If it is configurable it becomes an explicit effect. For example a `style: reformatted with new rules` should not require a new release imo.

ccopsey replied:
I see that you are following the spec.  But I feel there is a big difference between "should not" and "can not".  What if this was a continuous deployment scenario and every merge to main gets released and hence a version number?  Regardless, this PR has this covered.

hdevalke commented on src/cmd/version.rs:
```diff
@@ -94,11 +98,18 @@ impl VersionCommand {
+
+            let option_commit_type = types
+                .iter()
+                .find(|x| x.to_string() == commit.r#type.to_string());
```
```suggestion
                .find(|x| x.to_string() == commit.r#type);
```

hdevalke reviewed (changes requested):
don't forget to use a conventional commit message for your commits ;)

ccopsey commented:
> don't forget to use a conventional commit message for your commits ;)

Thanks for taking the time to review my proposal.  Do you prefer every commit to be conventional?  I was expecting a squash merge and hence the title of the PR would take contain the label.  Would you prefer me to reword my commit via rebase?  And hence break the history of the PR?

hdevalke commented:
> > don't forget to use a conventional commit message for your commits ;)
> 
> Thanks for taking the time to review my proposal. Do you prefer every commit to be conventional? I was expecting a squash merge and hence the title of the PR would take contain the label. Would you prefer me to reword my commit via rebase? And hence break the history of the PR?



> > don't forget to use a conventional commit message for your commits ;)
> 
> Thanks for taking the time to review my proposal. Do you prefer every commit to be conventional? I was expecting a squash merge and hence the title of the PR would take contain the label. Would you prefer me to reword my commit via rebase? And hence break the history of the PR?

Indeed I can use the squash and merge. I normally use the rebase and merge as the workflows will make sure a conventional commit is used for each commit.

ccopsey commented:
> Indeed I can use the squash and merge. I normally use the rebase and merge as the workflows will make sure a conventional commit is used for each commit.

Since the workflows failed I fixed the style issues and squashed the lot down to a single conventional commit.

hdevalke commented:
I fixed the remaining issues here: https://github.com/convco/convco/tree/BRANCH_NAME

hdevalke commented:
Merged with 607c9b17e32227334f142f94519c5a92e4155d4a
